File: api/app/schemas/pool_config.py
"""Pydantic discriminated union for pool rules_json validation.

This schema is the Python mirror of the Zod discriminated union in
packages/js-core/src/pool-config.ts.  Both must stay in sync.
"""

from __future__ import annotations

from typing import Annotated, Any, Literal, Union

from pydantic import BaseModel, ConfigDict, Field, TypeAdapter, ValidationError


class RvccPoolConfig(BaseModel):
    """Rules for RVCC variant: 7 picks, best 5 count, min 5 cuts, no buckets."""

    model_config = ConfigDict(extra="forbid")

    variant: Literal["rvcc"]
    pick_count: Literal[7] = 7
    count_best: Literal[5] = 5
    min_cuts_to_qualify: Literal[5] = 5
    uses_buckets: Literal[False] = False


class CrestmontPoolConfig(BaseModel):
    """Rules for Crestmont variant: 6 picks, best 4 count, min 4 cuts, 6 buckets."""

    model_config = ConfigDict(extra="forbid")

    variant: Literal["crestmont"]
    pick_count: Literal[6] = 6
    count_best: Literal[4] = 4
    min_cuts_to_qualify: Literal[4] = 4
    uses_buckets: Literal[True] = True
    bucket_count: Literal[6]  # required — no default enforces explicit declaration


PoolConfig = Annotated[
    Union[RvccPoolConfig, CrestmontPoolConfig],
    Field(discriminator="variant"),
]

_POOL_CONFIG_ADAPTER: TypeAdapter[PoolConfig] = TypeAdapter(PoolConfig)


class PoolConfigValidator:
    """Validates pool rules_json against the known pool variants.

    Acts as the single enforcement point so create and update paths
    share identical validation semantics.
    """

    @staticmethod
    def validate(rules_json: dict[str, Any]) -> None:
        """Validate rules_json against the discriminated union.

        Raises ValueError with a dotted field path prefix on failure so
        callers (and FastAPI field validators) can surface actionable errors.
        """
        try:
            _POOL_CONFIG_ADAPTER.validate_python(rules_json)
        except ValidationError as exc:
            errors = exc.errors(include_url=False)
            first = errors[0] if errors else {}
            loc_parts = [str(p) for p in first.get("loc", [])]
            loc = ".".join(loc_parts) if loc_parts else ""
            msg = first.get("msg", "Invalid pool configuration")
            field_path = f"rules_json.{loc}" if loc else "rules_json"
            raise ValueError(f"{field_path}: {msg}") from exc

File: packages/js-core/src/pool-config.ts
/**
 * Pool configuration Zod discriminated union.
 *
 * Mirrors api/app/schemas/pool_config.py — both must stay in sync.
 * Discriminator field: `variant`.
 */

import { z } from "zod";

export const RvccPoolConfigSchema = z.object({
  variant: z.literal("rvcc"),
  pick_count: z.literal(7).default(7),
  count_best: z.literal(5).default(5),
  min_cuts_to_qualify: z.literal(5).default(5),
  uses_buckets: z.literal(false).default(false),
});

export const CrestmontPoolConfigSchema = z.object({
  variant: z.literal("crestmont"),
  pick_count: z.literal(6).default(6),
  count_best: z.literal(4).default(4),
  min_cuts_to_qualify: z.literal(4).default(4),
  uses_buckets: z.literal(true).default(true),
  // bucket_count has no default — must be explicitly provided
  bucket_count: z.literal(6),
});

export const PoolConfigSchema = z.discriminatedUnion("variant", [
  RvccPoolConfigSchema,
  CrestmontPoolConfigSchema,
]);

export type RvccPoolConfig = z.infer<typeof RvccPoolConfigSchema>;
export type CrestmontPoolConfig = z.infer<typeof CrestmontPoolConfigSchema>;
export type PoolConfig = z.infer<typeof PoolConfigSchema>;

/**
 * Parse unknown input as a typed PoolConfig.
 * Throws ZodError on invalid input.
 */
export function parsePoolConfig(input: unknown): PoolConfig {
  return PoolConfigSchema.parse(input);
}
